### PR TITLE
BUG: Perform i8 conversion for datetimelike IntervalTree queries

### LINKED ISSUE #20636: BUG: indexing in datetime IntervalIndex with duplicate values fails
#### Code Sample
```python
import pandas as pd
start_end_idx = pd.IntervalIndex.from_arrays(start, end)
start_end_idx.name = 'port_time'
mm = mm.set_index(['imo','port_id', start_end_idx])
global_bins = mm.index
global_bins.get_loc((8801527, 5610.0, pd.to_datetime('2016-10-16 16:57:44.316')))
```
```
/anaconda3/envs/mariquant/lib/python3.6/site-packages/pandas/core/indexes/interval.py in _engine(self)
    283     @cache_readonly
    284     def _engine(self):
--> 285         return IntervalTree(self.left, self.right, closed=self.closed)
    286 
    287     @property

pandas/_libs/intervaltree.pxi in pandas._libs.interval.IntervalTree.__init__()

KeyError: ('datetime64[ns]', 'right')
```
#### Problem description
It is impossible to use non unique IntervalIndex with datetimes as start-end points. Both index.get_loc and DataFrame.loc produce the same error in IntervalTree.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

INSTALLED VERSIONS
------------------
commit: None

pandas: 0.22.0
pytest: None
pip: 9.0.1
setuptools: 38.5.1
Cython: None
numpy: 1.14.2
scipy: 1.0.0
pyarrow: None
xarray: 0.10.2
IPython: 6.2.1
sphinx: 1.7.1
patsy: 0.5.0
dateutil: 2.7.0
pytz: 2018.3
blosc: None
bottleneck: 1.2.1
tables: None
numexpr: None
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: 1.2.5
pymysql: None
psycopg2: 2.7.4 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by jschendel**
- [X] closes #20636
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

I've seen a few complaints about this on StackOverflow, so wanted to get this into 0.24.0 before I try implementing larger changes like the new behavior specs and interval accessor.

### CODE REVIEW

**Comment by mroeschke** [comment]
Could you also add tests for `Period`s as well? I think they also need i8 conversion (though possibly less commonly used).

**Comment by jschendel** [comment]
We currently do not allow an `IntervalIndex` to have `Period` endpoints, so I don't think tests are necessary:
```python
In [2]: pd.IntervalIndex.from_breaks(pd.period_range('2018Q1', freq='Q', periods=3))
---------------------------------------------------------------------------
ValueError: Period dtypes are not supported, use a PeriodIndex instead
```

Interestingly enough, it looks like an `Interval` _can_ have `Period` endpoints, which seems like a bug:
```python
In [3]: pd.Interval(pd.Period('2018Q1', freq='Q'), pd.Period('2018Q3', freq='Q'))
Out[3]: Interval(Period('2018Q1', 'Q-DEC'), Period('2018Q3', 'Q-DEC'), closed='right')
```

**Comment by mroeschke** [comment]
Gotcha. Are there plans for IntervalIndex to support Period? That might get tricky since Periods themselves are conceptually 1 label for an interval of time.

**Comment by jschendel** [comment]
> Are there plans for IntervalIndex to support Period? 

I'm not aware of any such plans.  I don't think the same type algorithms would apply; a `Period` is basically a special case of an `Interval`, so you'd essentially have intervals with interval endpoints (e.g. `Period('2018Q1')` is conceptually `Interval('2018-01-01', '2018-04-01', closed='left')` with some additional special properties.).  Haven't fully thought it through though.

There is an open issue to have `Period`/`PeriodIndex` inherit from `Interval`/`IntervalIndex` though, allowing one to use any of the interval methods on periods.  Not quite as straight forward as it might initially seem though, as we allow a `Period` to be outside the bounds of `Timestamp`, so just using `Timestamp` endpoints for the `Interval` representation wouldn't cut it.

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -514,6 +519,41 @@ def _maybe_cast_indexed(self, key):
 
         return key
 
+    def _maybe_convert_i8(self, key):
+        if isinstance(key, Interval):
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: can you document this a bit, and can you simplify at all?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (jschendel) [reply pandas/core/indexes/interval.py]: Done. Look better?

**Comment by jreback** [comment]
@mroeschke as indicated above, a ``Period`` doesn't make sense as an element of an ``IntervalIndex``, though I don't think we actually preclude it, maybe we should. However implemented ``Period`` as a sub-class of ``Intevral`` and ``PeriodIndex`` as a sub-class of ``IntervalIndex``, *does* make sense (though may not be straightforward to actually do this).

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jschendel

**Pull request merged**